python/ray/experimental/raysort/main.py

```diff
@@ -190,146 +208,270 @@ def _dummy_sort_and_partition(part: np.ndarray,
 
 
 @ray.remote
-def mapper(boundaries: List[int], mapper_id: PartId,
-           path: Path) -> List[ray.ObjectRef]:
+@tracing_utils.timeit("map")
+def mapper(args: Args, boundaries: List[int], path: Path) -> List[np.ndarray]:
     logging_utils.init()
-    task_id = f"M-{mapper_id} Mapper"
-    logging.info(f"{task_id} starting {args}")
+
     if args.skip_input:
-        block_size = int(np.ceil(args.input_part_size / args.num_reducers))
-        return [
-            ray.put(
-                np.frombuffer(np.random.bytes(block_size), dtype=np.uint8))
-            for _ in range(args.num_reducers)
-        ]
+        part = np.frombuffer(
+            np.random.bytes(args.input_part_size), dtype=np.uint8)
+    else:
+        part = _load_partition(path)
 
-    part = _load_partition(path)
     sort_fn = _dummy_sort_and_partition \
         if args.skip_sorting else sortlib.sort_and_partition
     blocks = sort_fn(part, boundaries)
-    logging.info(f"{task_id} saving to object store")
-    return [ray.put(part[offset:offset + size]) for offset, size in blocks]
+    return [part[offset:offset + size] for offset, size in blocks]
 
 
-def _dummy_merge(blocks: List[np.ndarray], _n: int) -> Iterable[memoryview]:
-    for block in blocks:
+def _dummy_merge(
+        num_blocks: int, _n: int,
+        get_block: Callable[[int, int], np.ndarray]) -> Iterable[np.ndarray]:
+    blocks = [((i, 0), get_block(i, 0)) for i in range(num_blocks)]
+    while len(blocks) > 0:
+        (m, d), block = blocks.pop(random.randrange(len(blocks)))
         yield block
+        d_ = d + 1
+        block = get_block(m, d_)
+        if block is None:
+            continue
+        blocks.append(((m, d_), block))
 
 
-@ray.remote
-def reducer(reducer_id: PartId, *blocks: List[ray.ObjectRef]) -> PartitionInfo:
-    logging_utils.init()
-    task_id = f"R-{reducer_id} Reducer"
-    logging.info(f"{task_id} starting")
-    blocks = [np.copy(ray.get(block)) for block in blocks]
+@ray.remote(num_cpus=0, resources={"worker": 1e-3})
+def _write_data(path, data):
+    with open(path, "ab") as fout:
+        fout.write(data)
+
+
+def _merge_impl(args: Args,
+                M: int,
+                pinfo: PartInfo,
+                get_block: Callable[[int, int], np.ndarray],
+                skip_output=False,
+                remote_writer=True):
     merge_fn = _dummy_merge if args.skip_sorting else sortlib.merge_partitions
-    merger = merge_fn(blocks, args.reducer_batch_num_records)
-    if args.skip_output:
+    merger = merge_fn(M, args.reducer_output_chunk, get_block)
+
+    if skip_output:
         for datachunk in merger:
             del datachunk
-        logging.info(f"{task_id} done")
-        return None
     else:
-        pinfo = _make_partition_info(reducer_id, "output")
-        with open(pinfo.path, "wb") as fout:
+        if not remote_writer:
```

Comment:
What's the purpose of the remote writer? Shouldn't we always be writing locally?

Reply:
"remote" as in a Ray remote function. Using placement groups it will always run on the same node. This is for pipelining - offloading the write to disk to another process.

We can do this serially now and it won't have much difference. In the real benchmark you need to write to EBS which would be a bit slower.

Reply:
I see, let's just write serially for now, and we can measure the potential gain for this later. I think it will be easier to understand and measure the performance if we write serially.